File: src/main/java/org/tasks/dialogs/ThemePickerDialog.java
package org.tasks.dialogs;

import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.Drawable;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.TextView;

import org.tasks.R;
import org.tasks.injection.ForApplication;
import org.tasks.preferences.Preferences;
import org.tasks.themes.Theme;
import org.tasks.themes.ThemeCache;

import javax.inject.Inject;

import static com.todoroo.andlib.utility.AndroidUtilities.atLeastJellybeanMR1;

public class ThemePickerDialog {

    public enum ColorPalette {THEMES, COLORS, ACCENTS, WIDGET_BACKGROUND}

    public interface ThemePickerCallback {
        void themePicked(ColorPalette palette, int index);

        void initiateThemePurchase();
    }

    private DialogBuilder dialogBuilder;
    private Context context;
    private Preferences preferences;
    private ThemeCache themeCache;
    private Theme theme;

    @Inject
    public ThemePickerDialog(DialogBuilder dialogBuilder, @ForApplication Context context, Preferences preferences,
                             Theme theme, ThemeCache themeCache) {
        this.dialogBuilder = dialogBuilder;
        this.context = context;
        this.preferences = preferences;
        this.theme = theme;
        this.themeCache = themeCache;
    }

    public AlertDialog createDialog(final ColorPalette palette, final ThemePickerCallback callback) {
        if (palette == ColorPalette.THEMES || palette == ColorPalette.WIDGET_BACKGROUND) {
            theme = theme.withBaseTheme(themeCache.getThemeBase(2));
        }

        final String[] themes = context.getResources().getStringArray(getNameRes(palette));

        final boolean purchasedThemes = preferences.hasPurchase(R.string.p_purchased_themes);

        final LayoutInflater inflater = theme.getLayoutInflater(context);
        ListAdapter adapter = new ArrayAdapter<String>(context, R.layout.color_selection_row, themes) {
            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                TextView textView = (TextView) (convertView == null
                        ? inflater.inflate(R.layout.color_selection_row, parent, false)
                        : convertView);
                Drawable original = context.getResources().getDrawable(purchasedThemes || position < 2
                        ? R.drawable.ic_lens_black_24dp
                        : R.drawable.ic_vpn_key_black_24dp);
                Drawable wrapped = DrawableCompat.wrap(original.mutate());
                DrawableCompat.setTint(wrapped, getDisplayColor(palette, position));
                if (atLeastJellybeanMR1()) {
                    textView.setCompoundDrawablesRelativeWithIntrinsicBounds(wrapped, null, null, null);
                } else {
                    textView.setCompoundDrawablesWithIntrinsicBounds(wrapped, null, null, null);
                }
                textView.setText(themes[position]);
                return textView;
            }
        };

        return dialogBuilder.newDialog(theme)
                .setAdapter(adapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                        if (purchasedThemes || which < 2) {
                            callback.themePicked(palette, which);
                        } else {
                            callback.initiateThemePurchase();
                        }
                    }
                })
                .show();
    }

    private int getNameRes(ColorPalette palette) {
        switch (palette) {
            case COLORS:
                return R.array.colors;
            case ACCENTS:
                return R.array.accents;
            case WIDGET_BACKGROUND:
                return R.array.widget_background;
            default:
                return R.array.themes;
        }
    }

    private int getDisplayColor(ColorPalette palette, int index) {
        switch (palette) {
            case COLORS:
                return themeCache.getThemeColor(index).getPrimaryColor();
            case ACCENTS:
                return themeCache.getThemeAccent(index).getAccentColor();
            case WIDGET_BACKGROUND:
                return themeCache.getWidgetTheme(index).getBackgroundColor();
            default:
                return themeCache.getThemeBase(index).getContentBackground();
        }
    }
}

File: app/src/main/java/org/tasks/dialogs/ThemePickerDialog.kt
package org.tasks.dialogs

import android.app.Activity.RESULT_CANCELED
import android.app.Activity.RESULT_OK
import android.app.Dialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import dagger.hilt.android.AndroidEntryPoint
import org.tasks.R
import org.tasks.billing.Inventory
import org.tasks.themes.ThemeAccent
import org.tasks.themes.ThemeBase
import org.tasks.themes.ThemeBase.EXTRA_THEME_OVERRIDE
import javax.inject.Inject

@AndroidEntryPoint
class ThemePickerDialog : DialogFragment() {

    companion object {
        const val EXTRA_SELECTED = "extra_selected"
        const val EXTRA_WIDGET = "extra_widget"

        fun newThemePickerDialog(
            target: Fragment,
            rc: Int,
            selected: Int,
            widget: Boolean = false
        ): ThemePickerDialog {
            val args = Bundle()
            args.putInt(EXTRA_SELECTED, selected)
            args.putBoolean(EXTRA_WIDGET, widget)
            val dialog = ThemePickerDialog()
            dialog.setTargetFragment(target, rc)
            dialog.arguments = args
            return dialog
        }
    }

    @Inject lateinit var inventory: Inventory
    @Inject lateinit var dialogBuilder: DialogBuilder
    @Inject lateinit var accent: ThemeAccent
    @Inject lateinit var themeBase: ThemeBase

    var adapter: ArrayAdapter<String>? = null
    var dialog: AlertDialog? = null
    var selected = -1
    var widget = false

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        selected = savedInstanceState?.getInt(EXTRA_SELECTED) ?: requireArguments().getInt(EXTRA_SELECTED)
        widget = arguments?.getBoolean(EXTRA_WIDGET) ?: false
        val themes = resources.getStringArray(
            if (widget) R.array.widget_themes else R.array.base_theme_names
        )

        adapter = object : ArrayAdapter<String>(requireActivity(), R.layout.simple_list_item_single_choice, themes) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val view = super.getView(position, convertView, parent)
                val textColor = if (isAvailable(position)) {
                    R.color.text_primary
                } else {
                    R.color.text_tertiary
                }
                val text: TextView = view.findViewById(R.id.text1)
                text.setTextColor(context.getColor(textColor))
                return view
            }
        }

        dialog = dialogBuilder.newDialog()
                .setTitle(R.string.theme)
                .setSingleChoiceItems(adapter, requireArguments().getInt(EXTRA_SELECTED)) { _, which ->
                    selected = which
                    if (available()) {
                        deliverResult()
                    } else {
                        updateButton()
                        activity?.intent?.putExtra(EXTRA_THEME_OVERRIDE, which)
                        Handler().post {
                            ThemeBase(which).setDefaultNightMode()
                            activity?.recreate()
                            activity?.overridePendingTransition(R.anim.fragment_fade_enter, R.anim.fragment_fade_exit)
                        }
                    }
                }
                .setPositiveButton(R.string.ok) { _, _ ->
                    deliverResult()
                }
                .show()
        updateButton()
        return dialog as Dialog
    }

    override fun onCancel(dialog: DialogInterface) {
        if (available()) {
            deliverResult()
        } else {
            targetFragment?.onActivityResult(targetRequestCode, RESULT_CANCELED, null)
        }
    }

    private fun deliverResult() {
        dialog?.dismiss()
        targetFragment?.onActivityResult(targetRequestCode, RESULT_OK, Intent().putExtra(EXTRA_SELECTED, selected))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putInt(EXTRA_SELECTED, selected)
    }

    private fun updateButton() {
        val stringRes = if (available()) {
            R.string.cancel
        } else {
            R.string.button_subscribe
        }

        dialog?.getButton(AlertDialog.BUTTON_POSITIVE)?.text = getString(stringRes)
    }

    private fun available() = isAvailable(selected)

    private fun isAvailable(index: Int) =
        inventory.purchasedThemes() || ThemeBase(index).isFree || widget
}